torchgeo/datasets/mmflood.py

```diff
@@ -0,0 +1,565 @@
+
+        paths['image'] = cast(list[str | None], image_files)
+        paths['mask'] = cast(list[str | None], mask_files)
+        paths['dem'] = cast(list[str | None], dem_files)
+        paths['hydro'] = hydro_files
+
+        return paths
+
+    def __getitem__(self, query: BoundingBox) -> dict[str, Tensor]:
+        """Retrieve image/mask and metadata indexed by query.
+
+        Args:
+            query: (minx, maxx, miny, maxy, mint, maxt) coordinates to index
+
+        Returns:
+            sample of image, mask and metadata at that index
+
+        Raises:
+            IndexError: if query is not found in the index
+        """
+        data = super().__getitem__(query)
+        missing_data = data['image'].isnan().any(dim=0)
+        # Set all pixel values of invalid areas to 0, all mask values to 255
+        data['image'][:, missing_data] = 0
+        data['mask'][missing_data] = self._ignore_index
+        return data
+
+    def _merge_dataset_indices(self) -> None:
+        """Create a new R-tree out of the individual indices from Sentinel-1, DEM and hydrography datasets."""
+        ds1, _ = self.datasets
+        # Use ds1 index
+        self.index = ds1.index
+
+    def _download(self) -> None:
+        """Download the dataset."""
+
+        def _check_and_download(filename: str, url: str) -> None:
+            path = os.path.join(self.root, filename)
+            if not os.path.exists(path):
+                md5 = self._md5[filename] if self.checksum else None
+                download_url(url, self.root, filename, md5)
+            return
+
+        filename = self.metadata['filename']
+        filepath = os.path.join(self.root, filename)
+        if not os.path.exists(filepath):
+            for idx in range(self._nparts):
+                part_file = f'activations.tar.{idx:03}.gz.part'
+                url = self.url + part_file
+
+                _check_and_download(part_file, url)
+
+        _check_and_download(
+            self.metadata['metadata_file'], self.url + self.metadata['metadata_file']
+        )
+
+    def _extract(self) -> None:
+        """Extract the dataset."""
+        filepath = os.path.join(self.root, self.metadata['filename'])
+        if str(filepath).endswith('.tar.gz'):
+            extract_archive(filepath)
+
+    def _verify(self) -> None:
+        """Verify the integrity of the dataset."""
+        dirpath = os.path.join(self.root, self.metadata['directory'])
+        metadata_filepath = os.path.join(self.root, self.metadata['metadata_file'])
+        # Check if both metadata file and directory exist
+        if os.path.isdir(dirpath) and os.path.isfile(metadata_filepath):
+            # Check pairings of all files
+            self._verify_tif_pairings()
+            return
+        if not self.download:
+            raise DatasetNotFoundError(self)
+        self._download()
+        self._merge_tar_files()
+        self._extract()
+
+    def _verify_tif_pairings(self) -> None:
```

Comment:
We don't normally go to this level of detail in `_verify`, this could be removed if you want to keep the class simple